ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java

```diff
@@ -711,18 +654,135 @@ private void compareCookies(Set<Cookie> expected, Set<javax.servlet.http.Cookie>
+  }
+
+  /**
+   * Does the content addition and read verification based on the arguments provided.
+   * @param digestAlgorithm the digest algorithm to use. Can be empty or {@code null} if digest checking is not
+   *                        required.
+   * @param content the complete content.
+   * @param httpContents {@code content} in parts and as {@link HttpContent}. Should contain all the data in
+   * {@code content}.
+   * @param numContentsToAddBeforeRead the number of {@link HttpContent} to add before making the
+   * {@link NettyRequest#readInto(AsyncWritableChannel, Callback)} call.
+   * @throws Exception
+   */
+  private void doContentAddAndReadTest(String digestAlgorithm, ByteBuffer content, List<HttpContent> httpContents,
+      int numContentsToAddBeforeRead)
+      throws Exception {
+    if (numContentsToAddBeforeRead > httpContents.size()) {
+      throw new IllegalArgumentException("numContentsToAddBeforeRead is more than the pieces of content available");
+    }
+
+    NettyRequest nettyRequest = createNettyRequest(HttpMethod.POST, "/", null);
+    byte[] wholeDigest = null;
+    if (digestAlgorithm != null && !digestAlgorithm.isEmpty()) {
+      MessageDigest digest = MessageDigest.getInstance(digestAlgorithm);
+      digest.update(content);
+      wholeDigest = digest.digest();
+      content.rewind();
+      nettyRequest.setDigestAlgorithm(digestAlgorithm);
+    }
+
+    int bytesToVerify = 0;
+    int addedCount = 0;
+    for (; addedCount < numContentsToAddBeforeRead; addedCount++) {
+      HttpContent httpContent = httpContents.get(addedCount);
+      bytesToVerify += httpContent.content().readableBytes();
+      nettyRequest.addContent(httpContent);
+      assertEquals("Reference count is not as expected", 2, httpContent.refCnt());
+    }
+    ByteBufferAsyncWritableChannel writeChannel = new ByteBufferAsyncWritableChannel();
+    ReadIntoCallback callback = new ReadIntoCallback();
+    Future<Long> future = nettyRequest.readInto(writeChannel, callback);
+    readAndVerify(bytesToVerify, writeChannel, content);
+
+    bytesToVerify = 0;
+    for (; addedCount < httpContents.size(); addedCount++) {
+      HttpContent httpContent = httpContents.get(addedCount);
+      bytesToVerify += httpContent.content().readableBytes();
+      nettyRequest.addContent(httpContent);
+      assertEquals("Reference count is not as expected", 2, httpContent.refCnt());
+    }
+    readAndVerify(bytesToVerify, writeChannel, content);
+    verifyRefCnts(httpContents);
+    writeChannel.close();
+    if (callback.exception != null) {
+      throw callback.exception;
+    }
+    long futureBytesRead = future.get();
+    assertEquals("Total bytes read does not match (callback)", content.limit(), callback.bytesRead);
+    assertEquals("Total bytes read does not match (future)", content.limit(), futureBytesRead);
+
+    // check twice to make sure the same digest is returned every time
+    for (int i = 0; i < 2; i++) {
```

Comment:
To make sure that the same digest is returned. Notice that the first call sets the digest and the second one simply returns what was set.